### PR TITLE
provide clients to change the grpc codecs or create their own

### PR DESCRIPTION
**Pull request by k0shk0sh**
Provide the possibility for clients  to pass the codec for grpcClient to use or create their own that could be deflate, snappy or custom one. 

- Merged decoder and encoder into one class.
- Gzip is used still by default for backward compatibility.
- Not relying on the encoding header anymore but rather on what the client defines as in gRPC client and server has to always align on specific codec beforehand and therefore relying on the supplied codec is sufficient enough. 
- Updated GrpcDispatcher to use Identity by default.

Addressing: https://github.com/square/wire/issues/1948 and https://github.com/square/wire/issues/1777

### CODE REVIEW

**Comment by swankjesse** [comment]
I don’t think we should make the codec a public API. I think it’s too low-level an abstraction; end users should just want to configure compression, and even that I think we can come up with better defaults.

Perhaps we can compress messages >= 1024 bytes by default, and let users configure something different if they prefer that.

**Comment by k0shk0sh** [comment]
Yeah, that was also my thought at first too, but looking at gRPC documentation regards the encoding, client and server can have a custom codec and by making this as public API it allows this customization. But I'm open to suggestions how to make this better.

File: wire-library/wire-grpc-mockwebserver/src/main/java/com/squareup/wire/mockwebserver/GrpcDispatcher.kt [hunk 1]
```diff
@@ -155,7 +156,7 @@ class GrpcDispatcher(
     val source = GrpcMessageSource(
         source = request.body,
         messageAdapter = protoAdapter,
-        grpcEncoding = request.headers["grpc-encoding"]
+        codec = GrpcCodec.IdentityGrpcCodec
```
Comment (oldergod) [inline wire-library/wire-grpc-mockwebserver/src/main/java/com/squareup/wire/mockwebserver/GrpcDispatcher.kt:159]: Why don't we read it anymore?

File: wire-library/wire-grpc-client/src/commonMain/kotlin/com/squareup/wire/GrpcCodec.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2019 Square Inc.
```
Comment (oldergod) [inline wire-library/wire-grpc-client/src/commonMain/kotlin/com/squareup/wire/GrpcCodec.kt:2]: 2021

File: wire-grpc-sample/client/src/main/java/com/squareup/wire/whiteboard/GrpcClientProvider.kt [hunk 3]
```diff
@@ -47,6 +48,10 @@ object GrpcClientProvider {
 
   val grpcClient = GrpcClient.Builder()
       .client(okHttpClient)
+      .apply {
+        // GzipGrpcCodec used by default, use Identity or define your own extending GrpcEncoder
```
Comment (oldergod) [inline wire-grpc-sample/client/src/main/java/com/squareup/wire/whiteboard/GrpcClientProvider.kt:52]: https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: wire-library/wire-grpc-mockwebserver/src/main/java/com/squareup/wire/mockwebserver/GrpcDispatcher.kt (hunk 1 above)
Comment (k0shk0sh) [reply wire-library/wire-grpc-mockwebserver/src/main/java/com/squareup/wire/mockwebserver/GrpcDispatcher.kt:159]: The client supplies the codec for decoding and encoding as mentioned in my description.

Comment (k0shk0sh) [reply wire-library/wire-grpc-mockwebserver/src/main/java/com/squareup/wire/mockwebserver/GrpcDispatcher.kt:159]: From gRPC [documentation](https://www.google.com/url?sa=t&source=web&rct=j&url=https://github.com/grpc/grpc-go/blob/master/Documentation/encoding.md&ved=2ahUKEwielYGrnZvwAhVQhf0HHTVYBb4QFjAAegQIBRAC&usg=AOvVaw0SlY9CscYcxEOqSyk2lG0K)

> Codecs, by definition, must be symmetric, so the same desired Codec should be registered in both client and server binaries.

**Comment by swankjesse** [comment]
We won’t add support for custom codecs until there are widely used ones that we can’t build in. It seems that’s unlikely as even Go doesn’t support the gzip codec.

**Comment by k0shk0sh** [comment]
@swankjesse alright, im fine with this decision, what do you recommend would be the right way to allow clients to select which codec to use? enums, sealed classes?

**Review by swankjesse — CHANGES_REQUESTED** [review]
I don’t think we want user-pluggable codecs. As far as I know the only widely-available codecs are identity and gzip, and Go’s implementation screwed up gzip by making it off by default.

**Comment by swankjesse** [comment]
Obsoleted by https://github.com/square/wire/pull/1956

**Pull request closed without merging**